File: runner-scripts/dubbing-engine/lib/speakers.js
/**
 * Stage 4 – Speakers (pyannote speaker diarization)
 *
 * Calls Python script to identify who spoke when.
 * Optional stage – if pyannote is not installed or config says disabled, skip.
 */
const path = require('path');
const fs = require('fs');
const utils = require('./utils');

function getPythonScriptPath() {
  return path.resolve(__dirname, '..', 'python', 'speakers.py');
}

async function speakers(workDir, manifest) {
  const inputAudio = path.join(workDir, 'audio_mono.wav');
  const outputFile = path.join(workDir, 'speakers.json');
  const diarizeEnabled = manifest.dubbing?.diarization_enabled !== false;

  utils.logStep('SPEAKERS', `Enabled: ${diarizeEnabled}`);

  if (!diarizeEnabled) {
    console.log('[SPEAKERS] Diarization disabled in config – skipping');
    const placeholder = { skipped: true, speakers: [{ label: 'SPEAKER_00', segments: [] }] };
    utils.writeJson(outputFile, placeholder);
    return placeholder;
  }

  if (!fs.existsSync(inputAudio)) {
    throw new Error('[SPEAKERS] Mono audio not found – did extract stage run?');
  }

  let pyannoteAvailable = false;
  try {
    await utils.runProcess(utils.getPython(), [
      '-c', 'import pyannote.audio; print("pyannote_ok")'
    ], { stdio: 'pipe', timeoutMs: 10000, logLabel: 'SPEAKERS' });
    pyannoteAvailable = true;
    console.log('[SPEAKERS] pyannote.audio detected');
  } catch {
    console.log('[SPEAKERS] pyannote.audio not installed – skipping diarization');
  }

  if (pyannoteAvailable) {
    const pythonScript = getPythonScriptPath();
    await utils.runPython(pythonScript, [
      '--input', inputAudio,
      '--output', outputFile,
    ], { logLabel: 'SPEAKERS', timeoutMs: 300000 });
  } else {
    const placeholder = {
      skipped: true,
      reason: 'pyannote.audio not installed',
      speakers: [{ label: 'SPEAKER_00', segments: [{ start: 0, end: utils.getVideoDuration(inputAudio) || 0 }] }],
    };
    utils.writeJson(outputFile, placeholder);
    console.log('[SPEAKERS] ⚠️ Placeholder speaker data written');
  }

  const result = fs.existsSync(outputFile) ? utils.readJson(outputFile) : null;
  if (result && !result.skipped) {
    console.log(`[SPEAKERS] ✅ Done – ${result.speakers?.length || 0} speakers identified`);
  } else {
    // Mark as fallback if it was skipped due to missing pyannote (not user-disabled)
    if (result && diarizeEnabled && result.skipped) {
      result.fallback = true;
    }
    console.log(`[SPEAKERS] ⏭️ Skipped (pyannote not available or disabled)${result?.fallback ? ' ⚠️ (fallback)' : ''}`);
  }
  return result;
}

module.exports = { speakers };

File: runner-scripts/dubbing-engine/python/speakers.py
#!/usr/bin/env python3
"""
Dubbing Engine – Stage 4: Speaker Diarization (pyannote.audio)

Usage:
    python speakers.py --input audio_mono.wav --output speakers.json
"""
import argparse
import json
import os
import sys


def main():
    parser = argparse.ArgumentParser(description="pyannote speaker diarization")
    parser.add_argument("--input", required=True, help="Input mono WAV file")
    parser.add_argument("--output", required=True, help="Output JSON path")
    args = parser.parse_args()

    input_file = args.input
    output_file = args.output

    if not os.path.exists(input_file):
        print(f"[SPEAKERS] Error: Input not found: {input_file}", file=sys.stderr)
        sys.exit(1)

    os.makedirs(os.path.dirname(output_file), exist_ok=True)

    print(f"[SPEAKERS] Running speaker diarization on: {input_file}")
    sys.stdout.flush()

    result = {
        "skipped": True,
        "reason": "pyannote.audio not available or failed",
        "speakers": [
            {
                "label": "SPEAKER_00",
                "segments": [{"start": 0, "end": 300}],
            }
        ],
    }

    try:
        from pyannote.audio import Pipeline
        import torch

        # Try loading the diarization pipeline
        # Uses pretrained model from huggingface
        print("[SPEAKERS] Loading pyannote pipeline...")
        sys.stdout.flush()

        pipeline = Pipeline.from_pretrained(
            "pyannote/speaker-diarization-3.1",
            use_auth_token=os.environ.get("HUGGINGFACE_TOKEN", None),
        )

        if torch.cuda.is_available():
            pipeline.to(torch.device("cuda"))

        print("[SPEAKERS] Running diarization...")
        sys.stdout.flush()

        diarization = pipeline(input_file)

        # Process results
        speakers = {}
        for turn, _, speaker in diarization.itertracks(yield_label=True):
            if speaker not in speakers:
                speakers[speaker] = []
            speakers[speaker].append({
                "start": round(turn.start, 3),
                "end": round(turn.end, 3),
            })

        speaker_list = []
        for label, segments in sorted(speakers.items()):
            speaker_list.append({
                "label": label,
                "segments": segments,
            })

        result = {
            "skipped": False,
            "engine": "pyannote/speaker-diarization-3.1",
            "speakers": speaker_list,
        }

        print(f"[SPEAKERS] Found {len(speaker_list)} speaker(s)")

    except ImportError as e:
        print(f"[SPEAKERS] pyannote.audio not installed: {e}")
    except Exception as e:
        print(f"[SPEAKERS] Error: {e}", file=sys.stderr)

    with open(output_file, "w", encoding="utf-8") as f:
        json.dump(result, f, indent=2, ensure_ascii=False)

    if not result.get("skipped", True):
        print(f"[SPEAKERS] Done - {len(result.get('speakers', []))} speaker(s)")


if __name__ == "__main__":
    main()
